Pr 2 of 2: Add Microarray Pathway Analysis - GSEA example

cbethell opened the pull request:
## Analysis Purpose

<!-- What new analysis issue(s) does your PR address? -->

This PR closes issue #284

## Pull Request Stage

<!-- Select one of these and delete the other to let the reviewer know what stage this analysis example is at -->

This is a **Refined PR** - needs review of details and polishing

<!-- In the case of the Refined PR, link the draft PR here -->

The draft PR relevant to this PR is #339.

## Strategy

<!-- What was your strategy for this new or edited analysis? -->
This PR iterates on open PR #345 by adding the GSEA and visualization steps to the microarray GSEA example analysis.

<!-- Is there anything you found out in implementation that required you to shift strategies from what was originally discussed? -->

I copied over the steps from draft PR #339, incorporating the relevant feedback from there.

Therefore, this PR includes:
- The implementation of `clusterProfiler::GSEA()` -- per [this comment on the relevant draft PR](https://github.com/AlexsLemonade/refinebio-examples/pull/339/files#r513799939), some research into the function led me to the current implementation in this PR as it seems that `fgseaMultilevel()` is the recommended approach to use and the default for `clusterProfiler::GSEA()` when the `nPerm` argument is not provided (info gathered partly from the [fgsea docs](https://bioconductor.org/packages/devel/bioc/manuals/fgsea/man/fgsea.pdf) and partly from the warning message below):

<img width="788" alt="Screen Shot 2020-10-29 at 1 09 40 PM" src="https://user-images.githubusercontent.com/43576623/97742535-cb2e9900-1aba-11eb-9ba3-b7d047df3d99.png">

Also probably worth noting, when attempting to run `fgseaSimple()` and/or `fgseaMultilevel()` (without using the clusterProfiler wrapper), 0 results were returned (could be an error in my implementation however).

- To answer the [question re the pvalue cutoff](https://github.com/AlexsLemonade/refinebio-examples/pull/339/files#r513798718) on the draft PR, the p-value cutoff (FDR in this case) is an arbitrary cutoff which is now equal to *0.05* in this PR based on what seems to be _standard_ practice according to various docs (including but not limited to [Toward a gold standard for benchmarking gene set enrichment analysis](https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bbz158/5722384) and [clusterProfiler: an R Package for Comparing Biological Themes Among Gene Clusters](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3339379/)).

- The decision to create the pre-ranked gene vector based on the t-statistic rather than log fold change was explored based on [this comment from the draft PR](https://github.com/AlexsLemonade/refinebio-examples/pull/339#discussion_r513817275) and eventually made based on what is recommended in [Discovering statistically significant pathways in expression profiling studies](https://www.pnas.org/content/pnas/102/38/13544.full.pdf) and the explanation from a [biostars forum](https://www.biostars.org/p/200715/) which says "Try to understand how they relate to the question you're interested in e.g. if you're most interested in effect size then the fold change is what you should use but if you're more interested in statistical significance then look for one of the statistics taking into consideration the assumptions they make e.g. t-test" as I believe we are encouraging users to look at statistical significance here.

- Visualizations of the GSEA results using `enrichplot::gseaplot()` was included in this PR as well as they were in the draft PR

- Note that I also set the seed in this PR as I noticed that the results were slightly changing and realized that I did indeed this important step.

## Concerns/Questions for reviewers:

<!-- Are there any items you would like the reviewer to take an extra close look? -->

Note that at line 317, I left a prompt to `START REVIEW HERE` as this is the point where PR 1 of 2 (#345) ends and the new content being added in this PR begins (I will remove the prompt once #345 is merged).

Overall, do the decisions included in this PR make sense?

<!-- Any results that don't quite make sense? -->

<!-- Anywhere that you are unsure about wording or how to guide the user? -->

Does the context re `GSEA()` (the goal of it and the implementation) make sense?
Do the prompts seem sufficient for users?

## Analysis Pull Request Check List (roughly in order):

### Content checks
* [x] All `{{BLANKS}}` have been replaced with the correct content.
* [x] [Sources are cited](https://github.com/AlexsLemonade/refinebio-examples/blob/master/CONTRIBUTING.md#citing-sources-in-text)
* [x] Seed is set (if applicable)

### Formatting Checks
* [x] Removed any manual numbering of sections.
* [x] Removed any instances of chunk naming.
* [x] Comments and documentation are up to date.
* [x] All links have been checked and are properly formatted.

### Add datasets to S3
* [x] Added [data and metadata files to S3](https://github.com/AlexsLemonade/refinebio-examples/blob/master/CONTRIBUTING.md#adding-datasets-to-the-S3-bucket).

### Docker/Snakemake rendering components
* [ ] Added the `.html` [link to the navigation bar](https://github.com/AlexsLemonade/refinebio-examples/blob/master/CONTRIBUTING.md#add-new-analyses-to-the-navbar). (Done in PR 1 of 2 #345)
* [x] Any not yet added packages needed for this analysis have been added to the Dockerfile and it successfully builds.

cansavvy commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -0,0 +1,493 @@
+### Determine our pre-ranked genes list
+
+The `GSEA()` function takes a pre-ranked (sorted) named vector of statistics, where the names in the vector are gene identifiers.
+This is step 1 -- gene-level statistics.
```
```suggestion
This is step 1 -- create a gene list with statistics that GSEA will rank by.
```

cansavvy commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
````diff
@@ -0,0 +1,493 @@
+
+```{r}
+# Let's create a named vector ranked based on the t-statistic values -- we
+# already sorted in the t-stastic values in descending order in the previous
````
```suggestion
# already sorted in the t-statistic values in descending order in the previous
```

cansavvy commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -0,0 +1,493 @@
+
+### Run GSEA using the `GSEA()` function
+
+The enrichment score (ES) for a pathway, a pathway-level statistic, is calculated using our gene-level statistics.
```
```suggestion
The enrichment score (ES) is a pathway-level statistic, calculated using our gene-level statistics.
```

cansavvy commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -0,0 +1,493 @@
+
+The enrichment score (ES) for a pathway, a pathway-level statistic, is calculated using our gene-level statistics.
+Genes are ranked from most highly positive to most highly negative and weighted according to their gene-level statistic.
+A running score is calculated by starting with the most highly ranked genes and increasing the score when a gene is in the pathway and decreasing the score when a gene is not in the pathway.
```
```suggestion
An ES is calculated by starting with the most highly ranked genes and increasing the score when a gene is in the pathway and decreasing the score when a gene is not in the pathway.
```
What does it mean to `start with the most highly ranked genes`?

cansavvy replied:
I'm wondering if some of this information about _how_ the scores are calculated would be better absorbed if it were given along side the enrichment plots. 

(But the overview of GSEA bits should still be kept here I think).

cbethell replied:
> What does it mean to start with the most highly ranked genes?

The most highly ranked genes here refer to the highest positive and highest negative ranked genes according the t-statistic (tried to clarify this in the latest updates, please let me know if it is still not clear).

> I'm wondering if some of this information about how the scores are calculated would be better absorbed if it were given along side the enrichment plots.

> (But the overview of GSEA bits should still be kept here I think).

I also tried to rearrange this section as you suggested here in the latest updates to this PR.

cansavvy commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -0,0 +1,493 @@
+The enrichment score (ES) for a pathway, a pathway-level statistic, is calculated using our gene-level statistics.
+Genes are ranked from most highly positive to most highly negative and weighted according to their gene-level statistic.
+A running score is calculated by starting with the most highly ranked genes and increasing the score when a gene is in the pathway and decreasing the score when a gene is not in the pathway.
+The ES is the maximum deviation from zero.
```
Are you saying that ESs can be negative or positive where 0 indicates randomness? 

I think you could move this sentence up above the 355 line. Tell us intuitively what the score means and then if needed, we can get into how that's calculated.

cansavvy commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -0,0 +1,493 @@
+Genes are ranked from most highly positive to most highly negative and weighted according to their gene-level statistic.
+A running score is calculated by starting with the most highly ranked genes and increasing the score when a gene is in the pathway and decreasing the score when a gene is not in the pathway.
+The ES is the maximum deviation from zero.
+Significance is assessed by generating a null distribution by sampling random gene sets of the same size and an FDR (false discovery rate) value is calculated to account for multiple hypothesis testing.
```
This tidbit might be better absorbed when we look at the pvalues at 386 - 387? What do you think?

cbethell replied:
Agreed 👍

cansavvy commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -0,0 +1,493 @@
+gsea_result_df %>%
+  # Combining the `dplyr::desc()` function with `dplyr::arrange()` sorts the
+  # values in descending order
+  dplyr::arrange(dplyr::desc(NES))
```
Might be a good place to use `top_n()`? We don't really need everything printed out per se. Up to you.

cansavvy commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
````diff
@@ -0,0 +1,493 @@
+)
+```
+
+Notice how the genes that are in the gene set, indicated by the black bars, tend to be on the left side of the graph indicating that they have positive gene-level scores.
````
I like this sentence, can we add a sentence like it for describing the bottom panel? Aka what's the red dashed line about (looks like its the peak or max deviation from 0) we should probably highlight that and what it means.

cansavvy commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
````diff
@@ -0,0 +1,493 @@
+
+```{r}
+ggplot2::ggsave(file.path(plots_dir, "GSE71270_gsea_enrich_positive_plot.png"),
+  plot = ggplot2::last_plot()
````
In case people run things out of order (which I think the type of user who are following these examples may be prone to do that) perhaps we should assign the plots and supply them directly to the `plot` instead of using `last_plot()`?

cansavvy reviewed (commented):
I think this looks great! I just have some wording suggestions. 

I floated an idea about moving some of the "how this is calculated" info close to where the plots are since I think the plots illustrate that well and visuals can be helpful for absorbing th ideas. Let me know what you think.

cbethell commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
````diff
@@ -0,0 +1,493 @@
+```
+
+### Run GSEA using the `GSEA()` function
+
````
Per @jaclyn-taroni's [comment on PR #345](https://github.com/AlexsLemonade/refinebio-examples/pull/345#discussion_r518109401), we will likely want to  move the explanation below up before the `gene ID conversion` section and reiterate here since we now need to explain some scientific decision making in the `gene ID conversion` section.

jaclyn-taroni replied:
I think we do want to move this up, possibly before we install the libraries or under `Purpose of this analysis`. I would probably collect some of the information from throughout what is being added here (this section, around line 280, around line 395) and write up something pretty comprehensive as an introduction to the method to start. Then I'd _consider_ paring back what is written alongside the code chunks to reiterate the most salient points.

cbethell commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -0,0 +1,493 @@
+## Perform gene set enrichment analysis (GSEA)
+
+The goal of GSEA is to determine whether the members of a set of genes are randomly distributed throughout the pre-ranked gene list or primarily found at the top or bottom of the list.
+Genes are ranked based on phenotype so we expect that sets related to a distinct phenotype will tend to show the latter distribution [@clusterProfiler-book].
```
I tried rephrasing this in the most recent changes, please let me know if it is still unclear.

cbethell commented:
@cansavvy, I believe I mostly addressed your review suggestions. Please let me know if anything is still not clear or if I misinterpreted any of your suggestions!

cansavvy commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -404,30 +414,35 @@ Let's look for the gene set with the highest positive NES.
-  # values in descending order
-  dplyr::arrange(dplyr::desc(NES))
+  # Use `top_n()` to return the top n observation using `NES` as the ordering variable
+  dplyr::top_n(n = 1, wt = NES)
```
Can print out more than 1 just to show a little bit of a preview? Not 100% necessary but might be nice.

cansavvy commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -404,30 +414,35 @@ Let's look for the gene set with the highest positive NES.
 Notice how the genes that are in the gene set, indicated by the black bars, tend to be on the left side of the graph indicating that they have positive gene-level scores.
+The red dashed line indicates the peak ES score (the ES is the maximum deviation from zero).
+An ES is calculated by starting with the most highly ranked genes (according to the gene-level t-statistic values) and increasing the score when a gene is in the pathway and decreasing the score when a gene is not in the pathway.
+In this case, the most highly positive enrichment scores are being displayed.
```
```suggestion
In this case, the most highly positive enrichment score's data are being displayed.
```
The graph is only associated with one NES right?

cansavvy commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -437,27 +452,34 @@ Let's look for the gene set with the highest negative NES.
-  dplyr::arrange(NES)
+  # Use `top_n()` to return the top n observation using `NES` as the ordering
+  # variable -- note the use of `dplyr::desc()` to sort the values in descending order
+  dplyr::top_n(n = 1, wt = dplyr::desc(NES))
```
You can use -1 for this purpose instead of `desc()` That's what the docs show: https://dplyr.tidyverse.org/reference/top_n.html

cansavvy commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
````diff
@@ -437,27 +452,34 @@ Let's look for the gene set with the highest negative NES.
 ```
 
-The gene set `HALLMARK_E2F_TARGETS` has a highly negative NES.
+The gene set `HALLMARK_E2F_TARGETS` has the highest negative NES.
````
```suggestion
The gene set `HALLMARK_E2F_TARGETS` has the most negative NES.
```

It didn't occur to me before but "highest negative" sounds weird. Should we change this or is it not as weird as I'm thinking it is?

cbethell replied:
Agreed, for consistency, I mimicked this language when speaking about the _most_ positive NES as well. Please let me know if you think I should change that back to _highest positive_.

cansavvy commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
````diff
@@ -437,27 +452,34 @@ Let's look for the gene set with the highest negative NES.
 ```
 
 This gene set shows the opposite pattern -- genes in the pathway tend to be on the right side of the graph.
+Again, the red dashed line here indicates the peak ES score (the ES is the maximum deviation from zero).
````
This would probably be
```suggestion
Again, the red dashed line here indicates the maximum deviation from zero, in other words, the most negative ES score.
```

cansavvy reviewed (commented):
This looks good. I mostly have wording comments and little comments.

cbethell commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -142,18 +142,11 @@ library(org.Dr.eg.db)
 We will read in the differential expression results we will download from online.
 These results are from a zebrafish microarray experiment we used for [differential expression analysis for two groups](https://alexslemonade.github.io/refinebio-examples/02-microarray/differential-expression_microarray_02_2-groups.html) using [`limma`](https://bioconductor.org/packages/release/bioc/html/limma.html) [@Ritchie2015].
-The table contains summary statistics including Ensembl gene IDs, t-statistic values, and adjusted p-values (FDR in this case).
+The table contains summary statistics including Ensembl gene IDs, t-statistic values for each group, and adjusted p-values (FDR in this case).
```
You're correct, this snuck back in from what was copied from training modules (got rid of it in the last commit).

cbethell commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
````diff
@@ -275,7 +282,8 @@ Let's see a preview of `dge_mapped_df`.
 ```
 
-Now let's check to see if we have any Entrez IDs that mapped to multiple Ensembl IDs.
+We do not want duplicated gene identifiers for the GSEA steps later as the `GSEA()` function will throw a warning that having duplicate gene names may produce unexpected results.
````
I attempted to address this comment in the last commit via some restructuring and rewording. Please let me know if what I pushed in the latest commits are suffice (I was not sure if I should go more in depth about the warning message since users will not actually see it here).

cansavvy commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -333,15 +336,169 @@ Looks like we were able to successfully get rid of the duplicate gene identifier
-_Addressed in upcoming PR_
+The goal of GSEA is to determine whether the members of a set of genes are randomly distributed throughout a pre-ranked gene list or primarily found at the top or bottom of the list.
+As mentioned earlier, hallmark gene sets summarize and represent specific well-defined biological states or processes and display coherent expression of genes.
+We therefore expect that sets will tend to be highly- or lowly-ranked or, put another way, found at the top or bottom of the pre-ranked gene list [@clusterProfiler-book].
```
This sentence and 339 seem too similarly worded to be helpful. I think repetition or reiteration of an important point is good, but it feels a little too echoey. Can you explain it a different way or probably just delete it

cansavvy commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -333,15 +336,169 @@ Looks like we were able to successfully get rid of the duplicate gene identifier
+
+This gene set shows the opposite pattern -- genes in the pathway tend to be on the right side of the graph.
+Again, the red dashed line here indicates the maximum deviation from zero, in other words, the most negative ES score.
+An ES is calculated by starting with the most highly ranked genes (according to the gene-level t-statistic values) and increasing the score when a gene is in the pathway and decreasing the score when a gene is not in the pathway.
```
The sentence here looks like its still describing positive ES. Feel free to change around this wording suggestion, but we need it to describe NES. 
```suggestion
Here the ES is calculated by starting with the most lowly ranked genes (according to the gene-level t-statistic values) and making the score more negative when a gene is in the pathway. 
```

cansavvy commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -333,15 +336,169 @@ Looks like we were able to successfully get rid of the duplicate gene identifier
+# need to sort the t-statistic values in descending order here
+t_vector <- filtered_dge_mapped_df$t
+names(t_vector) <- filtered_dge_mapped_df$entrez_id
+t_vector <- sort(t_vector, decreasing = TRUE)
```
```suggestion
# We need to sort the t-statistic values in descending order here
t_vector <- sort(t_vector, decreasing = TRUE)
```
And then delete the second half of the comment above. I think placing the comment with its corresponding code makes it easier for the user to parse out which step is which.

cansavvy reviewed (commented):
I have a couple wording requests but otherwise this looks good to me. I don't feel the need to review this again.

jaclyn-taroni commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -333,15 +336,169 @@ Looks like we were able to successfully get rid of the duplicate gene identifier
 
-_Addressed in upcoming PR_
+The goal of GSEA is to determine whether the members of a set of genes are randomly distributed throughout a pre-ranked gene list or primarily found at the top or bottom of the list meaning that they are highly- or lowly-ranked.
+As mentioned earlier, hallmark gene sets summarize and represent specific well-defined biological states or processes and display coherent expression of genes so we would expect the latter distribution. [@clusterProfiler-book].
```
I don't think we _expect_ the latter distribution. It depends on the experiment! We might not always expect a particular pathway to be perturbed between our two groups. The way we run GSEA here (and almost every other way folks use GSEA) is a comparison between groups.

cbethell replied:
Ahh okay that makes sense, will update this section to more accurately reflect what's going on here.

jaclyn-taroni commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -333,15 +336,169 @@ Looks like we were able to successfully get rid of the duplicate gene identifier
 ## Perform gene set enrichment analysis (GSEA)
 
-_Addressed in upcoming PR_
+The goal of GSEA is to determine whether the members of a set of genes are randomly distributed throughout a pre-ranked gene list or primarily found at the top or bottom of the list meaning that they are highly- or lowly-ranked.
```
Is this the goal of the analysis or is it a description of the approach we use in service of our goal?

cbethell replied:
Good question, this goal was inspired by the goal in the related GSEA article: https://www.pnas.org/content/102/43/15545 which states that
"The goal of GSEA is to determine whether members of a gene set S tend to occur toward the top (or bottom) of the list L, in which case the gene set is correlated with the phenotypic class distinction."

However, now that you point it out, I would classify this as a description of the approach we use in service of our goal. Will update appropriately 👍

jaclyn-taroni commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
````diff
@@ -333,15 +336,169 @@ Looks like we were able to successfully get rid of the duplicate gene identifier
+head(gsea_results@result)
+```
+
+Significance is assessed by generating a null distribution by sampling random gene sets of the same size and an FDR (false discovery rate) value is calculated to account for multiple hypothesis testing.
````
I _believe_ that this implementation permutes the gene labels and uses the same gene set? It's basically accomplishing the same thing – you get an distribution of ES values for a gene sets that are the same size as the _real_ gene set but the relationship between the gene set and the true ranking is "broken."  I would describe it the same way that the docs or book for clusterProfiler do. (I think this may come directly from training material and I made it a point to talk about there would be two different ways to accomplish the goal / I just wanted the point to land using whatever mental model worked best for someone.) I may be overthinking it. I can't tell.

cbethell replied:
Gotcha! I'll revisit the docs and be sure to describe this section in a way which is more closely aligned with what's in the docs relevant to the package version we are using.

jaclyn-taroni reviewed (commented):
This looks like a great start! I had a few comments about wording and rearranging. I haven't looked at the code itself carefully or in great detail, but I assume that was a large part of @cansavvy's review at this point.

jaclyn-taroni commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -333,15 +338,169 @@ Looks like we were able to successfully get rid of the duplicate gene identifier
+### Determine our pre-ranked genes list
+
+The `GSEA()` function takes a pre-ranked and sorted named vector of statistics, where the names in the vector are gene identifiers.
+This is step 1 -- create a gene list with statistics that GSEA will rank by.
```
I feel like this may be leftover from training material where we say more about steps? Or maybe the training material doesn't say anything about steps after 1 #itwasapilot.

jaclyn-taroni commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -333,15 +338,169 @@ Looks like we were able to successfully get rid of the duplicate gene identifier
+
+Notice how the genes that are in the gene set, indicated by the black bars, tend to be on the left side of the graph indicating that they have positive gene-level scores.
+The red dashed line indicates the peak ES score (the ES is the maximum deviation from zero).
+An ES is calculated by starting with the most highly ranked genes (according to the gene-level t-statistic values) and increasing the score when a gene is in the pathway and decreasing the score when a gene is not in the pathway.
```
I _believe_ this is the first time we talk about this which seems _odd_. There may be something that isn't in the diff/was added earlier that I'm missing.

cbethell replied:
I believe when I addressed one of the earlier review comments, I removed it from an earlier section but have now included it in the brief intro under the `Purpose of this analysis` section.

jaclyn-taroni commented on 02-microarray/pathway-analysis_microarray_03_gsea.Rmd:
```diff
@@ -333,15 +338,169 @@ Looks like we were able to successfully get rid of the duplicate gene identifier
+
+This gene set shows the opposite pattern -- genes in the pathway tend to be on the right side of the graph.
+Again, the red dashed line here indicates the maximum deviation from zero, in other words, the most negative ES score.
+Here the ES is calculated by starting with the most lowly ranked genes (according to the gene-level t-statistic values) and making the score more negative when a gene is in the pathway.
```
Do you have documentation to support this? I would imagine we wouldn't be able to tell which "end" to start with and by virtue of not having many of the genes at the beginning of the list you'd decrement the score a lot and therefore end up with a negative enrichment score.

cbethell replied:
Good question, this language was suggested in an earlier review comment but without the documentation to support it. I fixed the language in accordance with your comment above.

jaclyn-taroni reviewed (approved):
This is in good shape and I am going to approve this. There's one comment that needs to be addressed around the negative enrichment scores before we merge this. Since we have to do #349, I think we should anticipate that some of what is currently written here will get rearranged and that means we should re-evaluate the text a bit when we get there. I see no reason to hold this up for that since it's hitting `staging` (woo the system is working)! Side note: can you file an issue to track updating this notebook specifically please?

cbethell commented:
> This is in good shape and I am going to approve this. There's one comment that needs to be addressed around the negative enrichment scores before we merge this. Since we have to do #349, I think we should anticipate that some of what is currently written here will get rearranged and that means we should re-evaluate the text a bit when we get there. I see no reason to hold this up for that since it's hitting `staging` (woo the system is working)! Side note: can you file an issue to track updating this notebook specifically please?

Addressed your comments in the [last commit](https://github.com/AlexsLemonade/refinebio-examples/pull/347/commits/82a97d882c72cff2d3ed67ee90e132aa69a48cd6) @jaclyn-taroni and filed issue #354 to track the context updates that we will need to make to the notebook in this PR.

jaclyn-taroni commented:
I approved earlier, so this is fine to merge in my opinion! We will have to re-evaluate when we get to #354 (thanks for filing) to make sure our explanations of in this notebook are well-aligned with the intro, but this is good to go.
